wade-review.md

```diff
@@ -0,0 +1,596 @@
+
+```java
+static class MyProcessorSupplier implements ProcessorSupplier<String, GitHubPRInfo, String, GitHubPRStateCounter> { 
+	static final String STORE_KEY = "state-counter";
+   @Override
+   public Processor<String, GitHubPRInfo, String, GitHubPRStateCounter> get() {
+		return new Processor<>() {
+
+    private KeyValueStore<String, GitHubPRStateCounter> kvStore;
+                @Override
+                public void init(final ProcessorContext<String, GitHubPRStateCounter> context) {
+                    this.kvStore = context.getStateStore(STORE_NAME);
+
+                    context.schedule(Duration.ofSeconds(1), PunctuationType.STREAM_TIME, timestamp -> {
+                        GitHubPRStateCounter entry = kvStore.get(STORE_KEY);
+                        System.out.printf("Store value %s%n", entry);
+                        context.forward(new Record<>("pr", entry, timestamp));
+                    });
+	}
+              @Override
+                public void process(final Record<String, GitHubPRInfo> record) {
+
+                }
+            };
+        }
+}
+```
+
+Next, I’ll write the process method to  take in the events, mark them as open or closed, increment the count, and stash them in the state store.
+
+```java
+static class MyProcessorSupplier implements ProcessorSupplier<String, GitHubPRInfo, String, GitHubPRStateCounter> { 
+	static final String STORE_KEY = "state-counter";
+   @Override
+   public Processor<String, GitHubPRInfo, String, GitHubPRStateCounter> get() {
+		return new Processor<>() {
+
+    private KeyValueStore<String, GitHubPRStateCounter> kvStore;
+                @Override
+                public void init(final ProcessorContext<String, GitHubPRStateCounter> context) {
+                    this.kvStore = context.getStateStore(STORE_NAME);
+
+                    context.schedule(Duration.ofSeconds(1), PunctuationType.STREAM_TIME, timestamp -> {
+                        GitHubPRStateCounter entry = kvStore.get(STORE_KEY);
+                        System.out.printf("Store value %s%n", entry);
+                        context.forward(new Record<>("pr", entry, timestamp));
+                    });
+	}
+              @Override
+                public void process(final Record<String, GitHubPRInfo> record) {
+                    GitHubPRInfo prInfo = record.value();
+
+                }
+            };
+        }
+}
+```
+We’ll get the value of the record and start a chain of logic. 
+
+```java
+SHOULD I ZOOM IN HERE I WONDER
+                    if(!prInfo.state().equals(NO_STATE)) {
+                        GitHubPRStateCounter stateCounter = kvStore.get(STORE_KEY);
+                        if (stateCounter == null) {
+                            stateCounter = new GitHubPRStateCounter();
+                        }
+                        if (prInfo.state().equalsIgnoreCase("open")) {
```

Comment:
Yes, it's pretty large:

```
{
  "type": "PULL_REQUESTS",
  "createdAt": 1465029800000,
  "data": {
    "url": "https://api.github.com/repos/apache/kafka/pulls/1301",
    "id": 68474576,
    "node_id": "MDExOlB1bGxSZXF1ZXN0Njg0NzQ1NzY=",
    "html_url": "https://github.com/apache/kafka/pull/1301",
    "diff_url": "https://github.com/apache/kafka/pull/1301.diff",
    "patch_url": "https://github.com/apache/kafka/pull/1301.patch",
    "issue_url": "https://api.github.com/repos/apache/kafka/issues/1301",
    "number": 1301,
    "state": "closed",
    "locked": false,
    "title": "KAFKA-3645: Fix ConsumerGroupCommand and ConsumerOffsetChecker to correctly read endpoint info from ZK",
    "user": {
      "login": "arunmahadevan",
      "id": 6792890,
      "node_id": "MDQ6VXNlcjY3OTI4OTA=",
      "avatar_url": "https://avatars.githubusercontent.com/u/6792890?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/arunmahadevan",
      "html_url": "https://github.com/arunmahadevan",
      "followers_url": "https://api.github.com/users/arunmahadevan/followers",
      "following_url": "https://api.github.com/users/arunmahadevan/following{/other_user}",
      "gists_url": "https://api.github.com/users/arunmahadevan/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/arunmahadevan/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/arunmahadevan/subscriptions",
      "organizations_url": "https://api.github.com/users/arunmahadevan/orgs",
      "repos_url": "https://api.github.com/users/arunmahadevan/repos",
      "events_url": "https://api.github.com/users/arunmahadevan/events{/privacy}",
      "received_events_url": "https://api.github.com/users/arunmahadevan/received_events",
      "type": "User",
      "site_admin": false
    },
    "body": "The host and port entries under /brokers/ids/<bid> gets filled only for PLAINTEXT security protocol. For other protocols the host is null and the actual endpoint is under \"endpoints\". This causes NPE when running the consumer group and offset checker scripts in a kerberized env. By always reading the host and port values from the \"endpoint\", a more meaningful exception would be thrown rather than a NPE.\n",
    "created_at": 1462039442000,
    "updated_at": 1465029800000,
    "closed_at": 1465029782000,
    "merged_at": null,
    "merge_commit_sha": "a3cbcfc342e26cde9a0ef4d8eac2f70dece6409a",
    "assignee": null,
    "assignees": "",
    "requested_reviewers": "",
    "requested_teams": "",
    "labels": [],
    "milestone": null,
    "draft": false,
    "commits_url": "https://api.github.com/repos/apache/kafka/pulls/1301/commits",
    "review_comments_url": "https://api.github.com/repos/apache/kafka/pulls/1301/comments",
    "review_comment_url": "https://api.github.com/repos/apache/kafka/pulls/comments{/number}",
    "comments_url": "https://api.github.com/repos/apache/kafka/issues/1301/comments",
    "statuses_url": "https://api.github.com/repos/apache/kafka/statuses/3c337295ed6938206e6891732f2da85b581fdc30",
    "head": {
      "label": "arunmahadevan:cg_kerb_fix",
      "ref": "cg_kerb_fix",
      "sha": "3c337295ed6938206e6891732f2da85b581fdc30",
      "user": {
        "login": "arunmahadevan",
        "id": 6792890,
        "node_id": "MDQ6VXNlcjY3OTI4OTA=",
        "avatar_url": "https://avatars.githubusercontent.com/u/6792890?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/arunmahadevan",
        "html_url": "https://github.com/arunmahadevan",
        "followers_url": "https://api.github.com/users/arunmahadevan/followers",
        "following_url": "https://api.github.com/users/arunmahadevan/following{/other_user}",
        "gists_url": "https://api.github.com/users/arunmahadevan/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/arunmahadevan/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/arunmahadevan/subscriptions",
        "organizations_url": "https://api.github.com/users/arunmahadevan/orgs",
        "repos_url": "https://api.github.com/users/arunmahadevan/repos",
        "events_url": "https://api.github.com/users/arunmahadevan/events{/privacy}",
        "received_events_url": "https://api.github.com/users/arunmahadevan/received_events",
        "type": "User",
        "site_admin": false
      },
      "repo": {
        "id": 52784277,
        "node_id": "MDEwOlJlcG9zaXRvcnk1Mjc4NDI3Nw==",
        "name": "kafka",
        "full_name": "arunmahadevan/kafka",
        "private": false,
        "owner": {
          "login": "arunmahadevan",
          "id": 6792890,
          "node_id": "MDQ6VXNlcjY3OTI4OTA=",
          "avatar_url": "https://avatars.githubusercontent.com/u/6792890?v=4",
          "gravatar_id": "",
          "url": "https://api.github.com/users/arunmahadevan",
          "html_url": "https://github.com/arunmahadevan",
          "followers_url": "https://api.github.com/users/arunmahadevan/followers",
          "following_url": "https://api.github.com/users/arunmahadevan/following{/other_user}",
          "gists_url": "https://api.github.com/users/arunmahadevan/gists{/gist_id}",
          "starred_url": "https://api.github.com/users/arunmahadevan/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/arunmahadevan/subscriptions",
          "organizations_url": "https://api.github.com/users/arunmahadevan/orgs",
          "repos_url": "https://api.github.com/users/arunmahadevan/repos",
          "events_url": "https://api.github.com/users/arunmahadevan/events{/privacy}",
          "received_events_url": "https://api.github.com/users/arunmahadevan/received_events",
          "type": "User",
          "site_admin": false
        },
        "html_url": "https://github.com/arunmahadevan/kafka",
        "description": "Mirror of Apache Kafka",
        "fork": true,
        "url": "https://api.github.com/repos/arunmahadevan/kafka",
        "forks_url": "https://api.github.com/repos/arunmahadevan/kafka/forks",
        "keys_url": "https://api.github.com/repos/arunmahadevan/kafka/keys{/key_id}",
        "collaborators_url": "https://api.github.com/repos/arunmahadevan/kafka/collaborators{/collaborator}",
        "teams_url": "https://api.github.com/repos/arunmahadevan/kafka/teams",
        "hooks_url": "https://api.github.com/repos/arunmahadevan/kafka/hooks",
        "issue_events_url": "https://api.github.com/repos/arunmahadevan/kafka/issues/events{/number}",
        "events_url": "https://api.github.com/repos/arunmahadevan/kafka/events",
        "assignees_url": "https://api.github.com/repos/arunmahadevan/kafka/assignees{/user}",
        "branches_url": "https://api.github.com/repos/arunmahadevan/kafka/branches{/branch}",
        "tags_url": "https://api.github.com/repos/arunmahadevan/kafka/tags",
        "blobs_url": "https://api.github.com/repos/arunmahadevan/kafka/git/blobs{/sha}",
        "git_tags_url": "https://api.github.com/repos/arunmahadevan/kafka/git/tags{/sha}",
        "git_refs_url": "https://api.github.com/repos/arunmahadevan/kafka/git/refs{/sha}",
        "trees_url": "https://api.github.com/repos/arunmahadevan/kafka/git/trees{/sha}",
        "statuses_url": "https://api.github.com/repos/arunmahadevan/kafka/statuses/{sha}",
        "languages_url": "https://api.github.com/repos/arunmahadevan/kafka/languages",
        "stargazers_url": "https://api.github.com/repos/arunmahadevan/kafka/stargazers",
        "contributors_url": "https://api.github.com/repos/arunmahadevan/kafka/contributors",
        "subscribers_url": "https://api.github.com/repos/arunmahadevan/kafka/subscribers",
        "subscription_url": "https://api.github.com/repos/arunmahadevan/kafka/subscription",
        "commits_url": "https://api.github.com/repos/arunmahadevan/kafka/commits{/sha}",
        "git_commits_url": "https://api.github.com/repos/arunmahadevan/kafka/git/commits{/sha}",
        "comments_url": "https://api.github.com/repos/arunmahadevan/kafka/comments{/number}",
        "issue_comment_url": "https://api.github.com/repos/arunmahadevan/kafka/issues/comments{/number}",
        "contents_url": "https://api.github.com/repos/arunmahadevan/kafka/contents/{+path}",
        "compare_url": "https://api.github.com/repos/arunmahadevan/kafka/compare/{base}...{head}",
        "merges_url": "https://api.github.com/repos/arunmahadevan/kafka/merges",
        "archive_url": "https://api.github.com/repos/arunmahadevan/kafka/{archive_format}{/ref}",
        "downloads_url": "https://api.github.com/repos/arunmahadevan/kafka/downloads",
        "issues_url": "https://api.github.com/repos/arunmahadevan/kafka/issues{/number}",
        "pulls_url": "https://api.github.com/repos/arunmahadevan/kafka/pulls{/number}",
        "milestones_url": "https://api.github.com/repos/arunmahadevan/kafka/milestones{/number}",
        "notifications_url": "https://api.github.com/repos/arunmahadevan/kafka/notifications{?since,all,participating}",
        "labels_url": "https://api.github.com/repos/arunmahadevan/kafka/labels{/name}",
        "releases_url": "https://api.github.com/repos/arunmahadevan/kafka/releases{/id}",
        "deployments_url": "https://api.github.com/repos/arunmahadevan/kafka/deployments",
        "created_at": 1456743052000,
        "updated_at": 1456743056000,
        "pushed_at": 1491545921000,
        "git_url": "git://github.com/arunmahadevan/kafka.git",
        "ssh_url": "git@github.com:arunmahadevan/kafka.git",
        "clone_url": "https://github.com/arunmahadevan/kafka.git",
        "svn_url": "https://github.com/arunmahadevan/kafka",
        "homepage": null,
        "size": 45308,
        "stargazers_count": 0,
        "watchers_count": 0,
        "language": "Java",
        "has_issues": false,
        "has_projects": true,
        "has_downloads": true,
        "has_wiki": false,
        "has_pages": false,
        "has_discussions": false,
        "forks_count": 0,
        "mirror_url": null,
        "archived": false,
        "disabled": false,
        "open_issues_count": 0,
        "license": {
          "key": "apache-2.0",
          "name": "Apache License 2.0",
          "spdx_id": "Apache-2.0",
          "url": "https://api.github.com/licenses/apache-2.0",
          "node_id": "MDc6TGljZW5zZTI="
        },
        "allow_forking": true,
        "is_template": false,
        "web_commit_signoff_required": false,
        "topics": [],
        "visibility": "public",
        "forks": 0,
        "open_issues": 0,
        "watchers": 0,
        "default_branch": "trunk"
      }
    },
    "base": {
      "label": "apache:trunk",
      "ref": "trunk",
      "sha": "9791401a778fee5a165373b923bbe0ece70defbe",
      "user": {
        "login": "apache",
        "id": 47359,
        "node_id": "MDEyOk9yZ2FuaXphdGlvbjQ3MzU5",
        "avatar_url": "https://avatars.githubusercontent.com/u/47359?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/apache",
        "html_url": "https://github.com/apache",
        "followers_url": "https://api.github.com/users/apache/followers",
        "following_url": "https://api.github.com/users/apache/following{/other_user}",
        "gists_url": "https://api.github.com/users/apache/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/apache/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/apache/subscriptions",
        "organizations_url": "https://api.github.com/users/apache/orgs",
        "repos_url": "https://api.github.com/users/apache/repos",
        "events_url": "https://api.github.com/users/apache/events{/privacy}",
        "received_events_url": "https://api.github.com/users/apache/received_events",
        "type": "Organization",
        "site_admin": false
      },
      "repo": {
        "id": 2211243,
        "node_id": "MDEwOlJlcG9zaXRvcnkyMjExMjQz",
        "name": "kafka",
        "full_name": "apache/kafka",
        "private": false,
        "owner": {
          "login": "apache",
          "id": 47359,
          "node_id": "MDEyOk9yZ2FuaXphdGlvbjQ3MzU5",
          "avatar_url": "https://avatars.githubusercontent.com/u/47359?v=4",
          "gravatar_id": "",
          "url": "https://api.github.com/users/apache",
          "html_url": "https://github.com/apache",
          "followers_url": "https://api.github.com/users/apache/followers",
          "following_url": "https://api.github.com/users/apache/following{/other_user}",
          "gists_url": "https://api.github.com/users/apache/gists{/gist_id}",
          "starred_url": "https://api.github.com/users/apache/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/apache/subscriptions",
          "organizations_url": "https://api.github.com/users/apache/orgs",
          "repos_url": "https://api.github.com/users/apache/repos",
          "events_url": "https://api.github.com/users/apache/events{/privacy}",
          "received_events_url": "https://api.github.com/users/apache/received_events",
          "type": "Organization",
          "site_admin": false
        },
        "html_url": "https://github.com/apache/kafka",
        "description": "Mirror of Apache Kafka",
        "fork": false,
        "url": "https://api.github.com/repos/apache/kafka",
        "forks_url": "https://api.github.com/repos/apache/kafka/forks",
        "keys_url": "https://api.github.com/repos/apache/kafka/keys{/key_id}",
        "collaborators_url": "https://api.github.com/repos/apache/kafka/collaborators{/collaborator}",
        "teams_url": "https://api.github.com/repos/apache/kafka/teams",
        "hooks_url": "https://api.github.com/repos/apache/kafka/hooks",
        "issue_events_url": "https://api.github.com/repos/apache/kafka/issues/events{/number}",
        "events_url": "https://api.github.com/repos/apache/kafka/events",
        "assignees_url": "https://api.github.com/repos/apache/kafka/assignees{/user}",
        "branches_url": "https://api.github.com/repos/apache/kafka/branches{/branch}",
        "tags_url": "https://api.github.com/repos/apache/kafka/tags",
        "blobs_url": "https://api.github.com/repos/apache/kafka/git/blobs{/sha}",
        "git_tags_url": "https://api.github.com/repos/apache/kafka/git/tags{/sha}",
        "git_refs_url": "https://api.github.com/repos/apache/kafka/git/refs{/sha}",
        "trees_url": "https://api.github.com/repos/apache/kafka/git/trees{/sha}",
        "statuses_url": "https://api.github.com/repos/apache/kafka/statuses/{sha}",
        "languages_url": "https://api.github.com/repos/apache/kafka/languages",
        "stargazers_url": "https://api.github.com/repos/apache/kafka/stargazers",
        "contributors_url": "https://api.github.com/repos/apache/kafka/contributors",
        "subscribers_url": "https://api.github.com/repos/apache/kafka/subscribers",
        "subscription_url": "https://api.github.com/repos/apache/kafka/subscription",
        "commits_url": "https://api.github.com/repos/apache/kafka/commits{/sha}",
        "git_commits_url": "https://api.github.com/repos/apache/kafka/git/commits{/sha}",
        "comments_url": "https://api.github.com/repos/apache/kafka/comments{/number}",
        "issue_comment_url": "https://api.github.com/repos/apache/kafka/issues/comments{/number}",
        "contents_url": "https://api.github.com/repos/apache/kafka/contents/{+path}",
        "compare_url": "https://api.github.com/repos/apache/kafka/compare/{base}...{head}",
        "merges_url": "https://api.github.com/repos/apache/kafka/merges",
        "archive_url": "https://api.github.com/repos/apache/kafka/{archive_format}{/ref}",
        "downloads_url": "https://api.github.com/repos/apache/kafka/downloads",
        "issues_url": "https://api.github.com/repos/apache/kafka/issues{/number}",
        "pulls_url": "https://api.github.com/repos/apache/kafka/pulls{/number}",
        "milestones_url": "https://api.github.com/repos/apache/kafka/milestones{/number}",
        "notifications_url": "https://api.github.com/repos/apache/kafka/notifications{?since,all,participating}",
        "labels_url": "https://api.github.com/repos/apache/kafka/labels{/name}",
        "releases_url": "https://api.github.com/repos/apache/kafka/releases{/id}",
        "deployments_url": "https://api.github.com/repos/apache/kafka/deployments",
        "created_at": 1313431576000,
        "updated_at": 1712856824000,
        "pushed_at": 1712854783000,
        "git_url": "git://github.com/apache/kafka.git",
        "ssh_url": "git@github.com:apache/kafka.git",
        "clone_url": "https://github.com/apache/kafka.git",
        "svn_url": "https://github.com/apache/kafka",
        "homepage": null,
        "size": 181975,
        "stargazers_count": 27220,
        "watchers_count": 27220,
        "language": "Java",
        "has_issues": false,
        "has_projects": true,
        "has_downloads": true,
        "has_wiki": false,
        "has_pages": false,
        "has_discussions": false,
        "forks_count": 13431,
        "mirror_url": null,
        "archived": false,
        "disabled": false,
        "open_issues_count": 1087,
        "license": {
          "key": "apache-2.0",
          "name": "Apache License 2.0",
          "spdx_id": "Apache-2.0",
          "url": "https://api.github.com/licenses/apache-2.0",
          "node_id": "MDc6TGljZW5zZTI="
        },
        "allow_forking": true,
        "is_template": false,
        "web_commit_signoff_required": false,
        "topics": [
          "kafka",
          "scala"
        ],
        "visibility": "public",
        "forks": 13431,
        "open_issues": 1087,
        "watchers": 27220,
        "default_branch": "trunk"
      }
    },
    "_links": {
      "self": {
        "href": "https://api.github.com/repos/apache/kafka/pulls/1301"
      },
      "html": {
        "href": "https://github.com/apache/kafka/pull/1301"
      },
      "issue": {
        "href": "https://api.github.com/repos/apache/kafka/issues/1301"
      },
      "comments": {
        "href": "https://api.github.com/repos/apache/kafka/issues/1301/comments"
      },
      "review_comments": {
        "href": "https://api.github.com/repos/apache/kafka/pulls/1301/comments"
      },
      "review_comment": {
        "href": "https://api.github.com/repos/apache/kafka/pulls/comments{/number}"
      },
      "commits": {
        "href": "https://api.github.com/repos/apache/kafka/pulls/1301/commits"
      },
      "statuses": {
        "href": "https://api.github.com/repos/apache/kafka/statuses/3c337295ed6938206e6891732f2da85b581fdc30"
      }
    },
    "author_association": "CONTRIBUTOR",
    "auto_merge": null,
    "active_lock_reason": null
  },
  "id": "68474576"
}
```

I think I'll [add a bit to the script](https://docs.google.com/document/d/1C0U9B_UAVB1NVAa3fuCT2yF606TdHCt8Q5iBjen02Vk/edit?disco=AAABJ2-4-z4) to open a record up when I'm presenting them coming in CC.

Reply:
Uhh...Yeah, maybe not... That's huge.

What if you just had a snippet of the JSON that you could show. Omit the parts you don't care about. You can make a note of that in the video.

> Here you can see a rough outline of what the JSON looks like. There's a lot more to it than this, but I wanted to narrow in on just the areas we are interested in.